test/tests/functional/pbs_maintenance_reservations.py

```diff
@@ -172,21 +395,120 @@ def test_maintenance_reconfirm_resv(self):
-              'reserve_start': now + 3600,
-              'reserve_end': now + 3605}
+              'reserve_start': now + 60,
+              'reserve_end': now + 7200}
         r1 = Reservation(TEST_USER, attrs=a1)
         rid1 = self.server.submit(r1)
 
         exp_attr = {'reserve_state': (MATCH_RE, 'RESV_CONFIRMED|2'),
                     'resv_nodes': '(%s:ncpus=1)' % self.momA.shortname}
         self.server.expect(RESV, exp_attr, id=rid1)
 
-        a2 = {'reserve_start': now + 3602,
-              'reserve_end': now + 3604,
-              '--hosts': self.momA.shortname}
-        r2 = Reservation(TEST_USER, attrs=a2)
-        self.server.submit(r2)
+        start = self.now_plus_to_rsub_format(minutes=1)
+        end = self.now_plus_to_rsub_format(hours=2)
+
+        ret = self.submit_maintenance(TEST_USER, start, end,
+                                      [self.mom.shortname])
+        self.assertEquals(ret['rc'], 0)
+        rid2 = ret['out'][0].split(" ")[0]
 
         exp_attr = {'reserve_state': (MATCH_RE, 'RESV_CONFIRMED|2'),
                     'resv_nodes': '(%s:ncpus=1)' % self.momB.shortname}
         self.server.expect(RESV, exp_attr, id=rid1)
+
+        a = {'Resource_List.select': '1:ncpus=1',
+             'Resource_List.walltime': '1000',
+             'queue': rid1.split('.')[0]}
+        j = Job(TEST_USER, attrs=a)
+        j.set_sleep_time(990)
+        jid1 = self.server.submit(j)
+
+        a = {'Resource_List.select': '1:ncpus=1',
+             'Resource_List.walltime': '1000',
+             'queue': rid2.split('.')[0]}
+        j = Job(TEST_USER, attrs=a)
+        j.set_sleep_time(990)
+        jid2 = self.server.submit(j)
+
+        self.logger.info("Wait for reservations to start (2 minutes)")
+        time.sleep(120)
+
+        exp_attr = {'reserve_state': (MATCH_RE, 'RESV_RUNNING|5')}
+        self.server.expect(RESV, exp_attr, id=rid1)
+
+        exp_attr = {'reserve_state': (MATCH_RE, 'RESV_RUNNING|5')}
+        self.server.expect(RESV, exp_attr, id=rid2)
+
+        exp_attr = {'job_state': 'R',
+                    'exec_vnode': "(%s:ncpus=1)" % self.momB.shortname}
+        self.server.expect(JOB, exp_attr, id=jid1)
+
+        exp_attr = {'job_state': 'R',
+                    'exec_vnode': "(%s:ncpus=1)" % self.momA.shortname}
+        self.server.expect(JOB, exp_attr, id=jid2)
+
+    def test_maintenance_progressive_degrade_reservation(self):
+        """
+        Test if the reservation is partially degraded by overlapping
+        maintenance reservation at first, and then do full overlap.
+        Test if the job in fully overlapped reservation will not run.
+        Two moms (-p "servers=M1,moms=M1:M2") are needed for this test.
+        """
+        now = int(time.time())
+
+        self.server.manager(MGR_CMD_SET, SERVER,
+                            {'managers': '%s@*' % TEST_USER})
+        self.server.manager(MGR_CMD_SET, SERVER,
+                            {'scheduler_iteration': 3})
```

Comment:
Curious, why lower the scheduler_iteration?  The server should kick cycles when needed.

Reply:
The same occurrence above was unnecessary (removed). But it is needed in the last two tests where the reconfirmation is done. If the reservation is degraded by maintenance reservation then no scheduler cycle is kicked. It will wait for the regular cycle. AFAIS, this is the same behavior as if the vnode of confirmed reservation turns down. Reconfirmation will also wait for the regular cycle.

Reply:
Actually, only one occurrence is necessary. The rest is removed.